### PR TITLE
OHRM5X-163: Improve employee list API

### PR DESCRIPTION
**Pull request by RajithaKumara**

### CODE REVIEW

File: phpunit.xml [hunk 1]
```diff
@@ -34,6 +34,7 @@
     <testsuite name="Pim Tests">
       <directory>symfony/plugins/orangehrmPimPlugin/test/Service</directory>
       <directory>symfony/plugins/orangehrmPimPlugin/test/Dao</directory>
+      <directory>symfony/plugins/orangehrmPimPlugin/test/Api</directory>
```
Comment (samanthajayasinghe) [inline phpunit.xml:37]: I think you can simplify this 
`symfony/plugins/orangehrmPimPlugin/test/*`

File: symfony/plugins/orangehrmCorePlugin/Api/V2/RequestParams.php [hunk 2]
```diff
@@ -144,6 +146,27 @@ public function getArrayOrNull(string $type, string $key, ?array $default = null
         return $this->$type->get($key, $default);
     }
 
+    /**
+     * @param string $type
+     * @param string $key
+     * @param array|null $default
+     * @return Base64Attachment|null
+     */
+    public function getAttachmentOrNull(string $type, string $key, ?array $default = null): ?Base64Attachment
```
Comment (samanthajayasinghe) [inline symfony/plugins/orangehrmCorePlugin/Api/V2/RequestParams.php:155]: getAttachment() is fine

File: symfony/plugins/orangehrmCorePlugin/Api/V2/Serializer/ModelTrait.php [hunk 3]
```diff
@@ -71,20 +71,42 @@ public function toArray(): array
...
             foreach ($this->filter as $index => $attribute) {
+                $key = empty($this->attributeNames[$index]) ? $attribute :
+                    $this->attributeNames[$index];
+
                 if (is_array($attribute)) {
                     $value = $this->entity;
-                    foreach ($attribute as $func) {
+                    foreach ($attribute as $i => $func) {
                         if (!is_null($value)) {
                             $value = call_user_func([$value, $func]);
```
Comment (samanthajayasinghe) [inline symfony/plugins/orangehrmCorePlugin/Api/V2/Serializer/ModelTrait.php:81]: $func($value) may work php7.4 >

File: symfony/plugins/orangehrmCorePlugin/Api/V2/Serializer/ModelTrait.php [hunk 4]
```diff
@@ -71,20 +71,42 @@ public function toArray(): array
...
+                                    $collectionAttributeNames
+                                );
+                                $key = array_slice($key, 0, $i + 1);
+                                break;
+                            }
                         }
                     }
                 } else {
                     // Only work with camel cased get methods with particular attribute
                     $getMethodName = "get" . ucfirst($attribute);
```
Comment (samanthajayasinghe) [inline symfony/plugins/orangehrmCorePlugin/Api/V2/Serializer/ModelTrait.php:106]: why we don't use magic __get()

File: symfony/plugins/orangehrmCorePlugin/Api/V2/Validator/Rules/Base64Attachment.php [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+            return false;
+        }
+        if (!in_array($input['type'], $this->allowedTypes)) {
+            return false;
+        }
+
+        if (!is_null($this->fileNameMinLength) || !is_null($this->fileNameMaxLength)) {
+            if (is_string($input['name'])) {
+                if (function_exists('mb_strlen')) {
+                    $fileNameLength = (int)mb_strlen($input['name']);
```
Comment (samanthajayasinghe) [inline symfony/plugins/orangehrmCorePlugin/Api/V2/Validator/Rules/Base64Attachment.php:75]: we should abstract str_* to text classes

File: symfony/plugins/orangehrmCorePlugin/Api/V2/Validator/Rules/Base64Attachment.php [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+                }
+            } else {
+                return false;
+            }
+        }
+
+        $size = intval($input['size']);
+        if ($this->checkSizeStrictly) {
+            $content = base64_decode($input['base64']);
+            if (function_exists('mb_strlen')) {
```
Comment (samanthajayasinghe) [inline symfony/plugins/orangehrmCorePlugin/Api/V2/Validator/Rules/Base64Attachment.php:91]: this is the result of less abstraction 
class text {
public function  strLength(string $text): int {
 }
}

File: symfony/plugins/orangehrmCorePlugin/Authorization/Exception/AuthorizationException.php [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+class AuthorizationException extends Exception
+{
+    /**
+     * @param string $entityType
+     * @param string $method
+     * @return static
+     */
+    public static function entityNotSupported(string $entityType, string $method): self
+    {
+        return new self(sprintf('Entity `%s` not supported, @ `%s`', $entityType, $method));
```
Comment (samanthajayasinghe) [inline symfony/plugins/orangehrmCorePlugin/Authorization/Exception/AuthorizationException.php:33]: btw .. sprintf little costly function 
`Entity {$entityType} not supported, @ {$method} make more readable

File: phpunit.xml (hunk 1 above)
Comment (RajithaKumara) [reply phpunit.xml:37]: yes. once migrate old tests, we can go with this way.

File: symfony/plugins/orangehrmCorePlugin/Api/V2/RequestParams.php (hunk 2 above)
Comment (RajithaKumara) [reply symfony/plugins/orangehrmCorePlugin/Api/V2/RequestParams.php:155]: this is not required scenario, then we have to allow null

File: symfony/plugins/orangehrmCorePlugin/Api/V2/Serializer/ModelTrait.php (hunk 4 above)
Comment (RajithaKumara) [reply symfony/plugins/orangehrmCorePlugin/Api/V2/Serializer/ModelTrait.php:106]: That's good idea

File: symfony/plugins/orangehrmCorePlugin/Authorization/Exception/AuthorizationException.php (hunk 7 above)
Comment (RajithaKumara) [reply symfony/plugins/orangehrmCorePlugin/Authorization/Exception/AuthorizationException.php:33]: Sure. I'll do, but this is exception handling not calling frequenly

Comment (RajithaKumara) [reply symfony/plugins/orangehrmCorePlugin/Authorization/Exception/AuthorizationException.php]: Done

File: symfony/plugins/orangehrmCorePlugin/Api/V2/Validator/Rules/Base64Attachment.php (hunk 5 above)
Comment (RajithaKumara) [reply symfony/plugins/orangehrmCorePlugin/Api/V2/Validator/Rules/Base64Attachment.php]: Done

File: symfony/plugins/orangehrmCorePlugin/Api/V2/Serializer/ModelTrait.php (hunk 3 above)
Comment (RajithaKumara) [reply symfony/plugins/orangehrmCorePlugin/Api/V2/Serializer/ModelTrait.php]: Done

**Pull request merged**